pypeit/core/skysub.py

```diff
@@ -808,10 +850,14 @@ def local_skysub_extract(sciimg, sciivar, tilts, waveimg, global_sky, rn2_img,
                 #  eqn. below is not valid. However, co-adds already have the model noise propagated correctly in sciivar,
                 #  so no need to re-model the variance.
                 if model_noise:
-                    var = np.abs(sky_bmodel + obj_bmodel - np.sqrt(2.0) * np.sqrt(rn2_img.flat[isub])) + rn2_img.flat[isub]
-                    var = var + adderr**2*(np.abs(sky_bmodel + obj_bmodel))**2
-                    modelivar.flat[isub] = (var > 0.0) / (var + (var == 0.0))
-                    #varnoobj.flat[isub] = var_no
+                    _proc_var = None if proc_var is None else proc_var.flat[isub]
+                    _count_scale = None if count_scale is None else count_scale.flat[isub]
+                    # NOTE: darkcurr must be a float for the call below to work.
+                    var = procimg.variance_model(rn2_img.flat[isub], counts=sky_bmodel+obj_bmodel,
```

Comment:
So sky_bmodel and obj_bmodel here are operating on an image that has already had all the "relative throughput" factors divided (multiplied?). As such, I see no reason for all of these arguments to be required for this routine. Basically this routine just needs two arguments from the image processing as far as I understand (I could be wrong), which is, 
following the notation in your docs: 

var_additive = s^2*(D*t_exp + V_rn + V_proc)
epsilon 

what is also confusing me here is that you appear to be dividing counts by count_scale in procimg.variance model, which is definitely incorrect for (sky_bmodel+obj_bmodel) since I'm already modeling the flattened counts. 

This is admittedly a little bit complicated given all the factors etc. and I may have understood something so let me know if you want to chat.

Reply:
This is a good suggestion, and I took it on.  The full variance model is now done in two steps, see `pypeit.images.rawimage.RawImage.build_ivar`.  The first step is to calculate the "base-level" variance, which you refer to as `var_additive` above.  Now the skysub and extraction routines just use this instead of requiring the readnoise, darkcurrent, exposure-time, and processing variance arguments.

The functions still need the scaling term because you need it to calculate the noise due to photon shot-noise.  I've changed how this is applied so that it's a bit more clear.  The reason its required is the following: `c = s*C` is the flat-fielded flux, where `C` is the number of electrons counted by the detector.  So the variance in `c` is `V_c = s^2 * V_C`.  Then, `V_C = C`, such that `V_c = s * c`.  I.e., unlike the counts recorded by the detector, the variance in the flat-fielded flux is `s*c`, not just `c`.  Hopefully the code is more clear now.